src/main/drivers/motor.c

```diff
@@ -0,0 +1,251 @@
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: jflyper
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "platform.h"
+
+#ifdef USE_MOTOR
+
+#include "common/maths.h"
+
+#include "config/feature.h"
+
+#include "drivers/dshot.h" // for DSHOT_ constants in initEscEndpoints; may be gone in the future
+
+#include "drivers/motor.h"
+#include "drivers/pwm_output.h" // for PWM_TYPE_* and others
+#include "drivers/time.h"
+#include "drivers/dshot_dpwm.h"
+
+#include "fc/rc_controls.h" // for flight3DConfig_t
+
+#include "pg/motor.h"
+
+static FAST_RAM_ZERO_INIT motorDevice_t *motorDevice;
+
+void motorShutdown(void)
+{
+    motorDevice->vTable.shutdown();
+    motorDevice->enabled = false;
+    motorDevice->initialized = false;
+    delayMicroseconds(1500);
+}
+
+void motorWriteAll(float *values)
+{
+#ifdef USE_PWM_OUTPUT
+    if (motorDevice->enabled) {
+#if defined(USE_DSHOT) && defined(USE_DSHOT_TELEMETRY)
+        if (!motorDevice->vTable.updateStart()) {
+            return;
+        }
+#endif
+        for (int i = 0; i < motorDevice->count; i++) {
+            motorDevice->vTable.write(i, values[i]);
+        }
+        motorDevice->vTable.updateComplete();
+    }
+#endif
+}
+
+int motorCount(void)
+{
+    return motorDevice->count;
+}
+
+// This is not motor generic anymore; should be moved to analog pwm module
+static void analogInitEndpoints(float outputLimit, float *outputLow, float *outputHigh, float *disarm, float *deadbandMotor3dHigh, float *deadbandMotor3dLow) {
+    if (featureIsEnabled(FEATURE_3D)) {
+        float outputLimitOffset = (flight3DConfig()->limit3d_high - flight3DConfig()->limit3d_low) * (1 - outputLimit) / 2;
+        *disarm = flight3DConfig()->neutral3d;
+        *outputLow = flight3DConfig()->limit3d_low + outputLimitOffset;
+        *outputHigh = flight3DConfig()->limit3d_high - outputLimitOffset;
+        *deadbandMotor3dHigh = flight3DConfig()->deadband3d_high;
+        *deadbandMotor3dLow = flight3DConfig()->deadband3d_low;
+    } else {
+        *disarm = motorConfig()->mincommand;
+        *outputLow = motorConfig()->minthrottle;
+        *outputHigh = motorConfig()->maxthrottle - ((motorConfig()->maxthrottle - motorConfig()->minthrottle) * (1 - outputLimit));
+    }
+}
+
+// End point initialization is called from mixerInit before motorDevInit; can't use vtable...
+void motorInitEndpoints(float outputLimit, float *outputLow, float *outputHigh, float *disarm, float *deadbandMotor3dHigh, float *deadbandMotor3dLow)
```

Comment:
My sense of logic tells me most of these functions should really be `motors...`. Not sure what the best thing is here.

Reply:
You mean the naming?

In my sense, if we look at the subject as individual motors, then it would be `motorsXXX`, while if we look at the subject as a layer function, then it would be `motorXXX`.

Reply:
Yeah, I guess you can think of it as the motor layer - some things like `motorCount()` do not make much sense then though.